drivers/gpio/gpio_eos_s3.c

```diff
@@ -0,0 +1,403 @@
+
+	return 0;
+}
+
+static int gpio_eos_s3_port_set_bits_raw(const struct device *dev,
+					 uint32_t mask)
+{
+	ARG_UNUSED(dev);
+
+	MISC_CTRL->IO_OUTPUT |= (mask & GPIOS_MASK);
+
+	return 0;
+}
+
+static int gpio_eos_s3_port_clear_bits_raw(const struct device *dev,
+					   uint32_t mask)
+{
+	ARG_UNUSED(dev);
+
+	MISC_CTRL->IO_OUTPUT &= ~(mask & GPIOS_MASK);
+
+	return 0;
+}
+
+static int gpio_eos_s3_port_toggle_bits(const struct device *dev,
+					uint32_t mask)
+{
+	ARG_UNUSED(dev);
+	uint32_t target_value;
+	uint32_t output_states = MISC_CTRL->IO_OUTPUT;
+
+	target_value = output_states ^ mask;
+	MISC_CTRL->IO_OUTPUT = (target_value & GPIOS_MASK);
+
+	return 0;
+}
+
+static int gpio_eos_s3_manage_callback(const struct device *dev,
+				       struct gpio_callback *callback, bool set)
+{
+	struct gpio_eos_s3_data *data = dev->data;
+
+	return gpio_manage_callback(&data->callbacks, callback, set);
+}
+
+static int gpio_eos_s3_pin_interrupt_configure(const struct device *dev,
+					       gpio_pin_t gpio_num,
+					       enum gpio_int_mode mode,
+					       enum gpio_int_trig trig)
+{
+	struct gpio_eos_s3_data *data = dev->data;
+	GPIOCfgTypeDef gpio_cfg;
+
+	gpio_cfg.ucGpioNum = gpio_num;
+	gpio_cfg.xPadConf = gpio_eos_s3_pad_select(dev, gpio_num);
+
+	if (mode == GPIO_INT_MODE_DISABLED) {
+		/* Get IRQ number which should be disabled */
+		int irq_num = gpio_eos_s3_get_irq_num(gpio_cfg.xPadConf->ucPin);
+
+		if (irq_num < 0) {
+			return -EINVAL;
+		}
+
+		/* Disable IRQ */
+		INTR_CTRL->GPIO_INTR_EN_M4 &= ~BIT((uint32_t)irq_num);
+
+		/* Clear IRQ mask and corresponding IRQ number */
+		data->gpio_irq_mask &= ~BIT((uint32_t)irq_num);
+		data->gpio_irqs[irq_num] = DISABLED_GPIO_IRQ;
+
+		/* Clear configuration */
+		INTR_CTRL->GPIO_INTR_TYPE &= ~((uint32_t)(BIT(irq_num)));
+		INTR_CTRL->GPIO_INTR_POL &= ~((uint32_t)(BIT(irq_num)));
+	} else {
+		/* Prepare configuration */
+		if (mode == GPIO_INT_MODE_LEVEL) {
+			gpio_cfg.intr_type = LEVEL_TRIGGERED;
+			if (trig == GPIO_INT_TRIG_LOW) {
+				gpio_cfg.pol_type = FALL_LOW;
+			} else {
+				gpio_cfg.pol_type = RISE_HIGH;
+			}
+		} else {
+			gpio_cfg.intr_type = EDGE_TRIGGERED;
+			switch (trig) {
+			case GPIO_INT_TRIG_LOW:
+				gpio_cfg.pol_type = FALL_LOW;
+				break;
+			case GPIO_INT_TRIG_HIGH:
+				gpio_cfg.pol_type = RISE_HIGH;
+				break;
+			case GPIO_INT_TRIG_BOTH:
+				return -ENOTSUP;
+			}
+		}
+
+		/* Set IRQ configuration */
+		int irq_num = HAL_GPIO_IntrCfg(&gpio_cfg);
+
+		if (irq_num < 0) {
+			return -EINVAL;
+		}
+
+		data->gpio_irqs[irq_num] = gpio_num;
+
+		data->gpio_irq_mask |= BIT((uint32_t)irq_num);
+
+		/* Enable IRQ */
+		INTR_CTRL->GPIO_INTR_EN_M4 |= BIT((uint32_t)irq_num);
+	}
+
+	return 0;
+}
+
+static void gpio_eos_s3_isr(const struct device *dev)
+{
+	struct gpio_eos_s3_data *data = dev->data;
+	uint32_t intr_status = INTR_CTRL->GPIO_INTR;
+	uint32_t enabled_int = intr_status & data->gpio_irq_mask;
+
+	INTR_CTRL->GPIO_INTR |= enabled_int;
```

Comment:
Again, I didn't find detailed information about interrupt controller in the datasheet. It may work differently than I assume. However, it's an error for the interrupt routine not to clear all of the active interrupt sources. If we don't clear the status bit we will enter the interrupt handler as soon as we exit it. Effectively it works as an endless loop. I propose to write it as
```
	INTR_CTRL->GPIO_INTR |= intr_status;
```
You may add an `__ASSERT` to verify that there are no bugs in the driver code and there was no other interrupt than that allowed by `data->gpio_irq_mask`.

Reply:
We should clear the interrupt statuses as you suggested. In the result I dropped the `gpio_irq_mask` because the control of fired callbacks is done already with the implemented condition: `data->gpio_irqs[irq_num] != DISABLED_GPIO_IRQ`.

I've tested the changes and all tests passed.